### ISSUE TITLE
Add support for tls-sni-01

### ISSUE THREAD

**Issue by bbaugh** [Neilpang/acme.sh#61]
I found some documentation for how us the tls-sni-01 challenge support to a client:

https://github.com/ietf-wg-acme/acme/blob/master/draft-ietf-acme-acme.md

In general simply changing the VTYPE_HTTP to "tls-sni-01" does all the proper steps but fails to verify with a poorly worded error. My understanding is that if the script where to add the proper subjectAlternativeName to a self signed cert and asking the user to install it before proceeding then it SHOULD world.

**Comment by Neilpang** [comment]
I  knew that.

but it's rarely  used and with much complexity.

Using either http-01 or dns  is enough, I think.

**Comment by bbaugh** [comment]
Thank you for your work on this script it is quite easy to use and read.

Many ISPs do not allow port 80 but do allow 443 and dns only works with full domains (e.g. not usable with free subdomains). Basically there are clear use cases for tls-sni-01 that cannot be met with  http-01 or dns.

**Comment by Neilpang** [comment]
interesting.
keep this issue open. we will support it later.

**Comment by Neilpang** [comment]
And I think  dns-01  is also usable with any subdomains.  you can try:

```
le  issue  dns    sub.domain.com
```

**Comment by bbaugh** [comment]
Sorry I should have been more explicit: dns doesn't work with sub-domains for which you are not the owner. That is free subdomains obtainable via http://freedns.afraid.org or the like do not allow one to add TXT entries to the DNS record which start with '_' making it impossible to use the DNS verification provided by Let's Encrypt. So it is an issue with free sub-domains and Let's Encrypt's interaction NOT this script which works fine for all dns challenges for which a user can act.

**Comment by Neilpang** [comment]
Oh,  I understand.

I remember, you are allowed to specify a third-party  dns server of the free domains from http://freedns.afraid.org 

Am I wrong ?

**Comment by bbaugh** [comment]
That is correct. So you can have 'cowsgo.mooo.com' Where you only have control off the 'cowsgo' portion.

**Comment by Neilpang** [comment]
I just registered a free subdomain:  myletest2.mooo.com

Then add a `NS` record of this subdomain, pointing to third dns server:  https://cloudns.net

Then I was managed to add `	_acme-challenge.myletest2.mooo.com`  there.

The certificate is almost done:
```
...
Verifying:myletest2.mooo.com
Success
Skip for removelevel:
Verify finished, start to sign.
Sign failed: "detail":"Error creating new cert :: Too many certificates already issued for: mooo.com"
...

```

Which means that  the root domain mooo.com  is not in public list. so, I'm afraid you can not issue cert for the free domains.

More details:  https://community.letsencrypt.org/t/too-many-certificates-already-issued-for-myfritz-net/10416

**Comment by fluffymadness** [comment]
tls-sni-01 would be nice :), because some isp's block port 80 but not 443.
If you don't want to register a domain and just want to use dyndns...then you have a problem.

**Comment by bbaugh** [comment]
The "public list" does contain many of the domains offered by freedns.afraid.org just not all. I just used mooo.com as an example.

**Comment by nihilus** [comment]
Any progress on this? Like @bbaugh say: many ISP's block port 80 and allows port 443.

**Comment by Neilpang** [comment]
Hi  @nihilus 

I just added tls support in a new branch `tls`

https://github.com/Neilpang/acme.sh/tree/tls

usage:

```
acme.sh  --issue  --tls   -d  aaa.com
```

Please have a try.

I will merge it to branch soon.

Thanks.

**Comment by nihilus** [comment]
@neilpang that is awesome. You rock. This will definitely help our customers enabling SSL if their ISPs block port 80.

**Comment by Neilpang** [comment]
Merged to master branch:  e22bcf7cb4e7c49226f9129b226c763874d0e324

Thanks.
